README
                     ====================================
                       Package "Date::Calc" Version 4.0
                     ====================================


This package is available for download either from my web site at

                  http://www.engelschall.com/u/sb/download/

or from any CPAN (= "Comprehensive Perl Archive Network") mirror server:

                  http://www.perl.com/CPAN/authors/id/STBEY/

The package consists of a C library (useful for C developers) which is the
core of a Perl (wrapper) module (for easy access to the library from Perl).

The C library is specifically designed so that it can be used stand-alone,
without Perl.


Legal issues:
-------------

This package with all its parts is

Copyright (c) 1995, 1996, 1997, 1998 by Steffen Beyer.
All rights reserved.

This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself, i.e., under the
terms of the "Artistic License" or the "GNU General Public License".

The C library at the core of this Perl module can additionally
be redistributed and/or modified under the terms of the "GNU
Library General Public License".

Please refer to the files "Artistic", "GNU_GPL" and "GNU_LGPL"
in this distribution for details!


Prerequisites:
--------------

Perl version 5.000 or higher, and an ANSI C compiler (!)
                                     ======

Installation:
-------------

Please see the file "INSTALL" in this distribution for instructions on
how to install this package.


Changes over previous versions:
-------------------------------

Please refer to the file "CHANGES" in this distribution for a detailed
version history and instructions on how to upgrade existing applications.


What does it do:
----------------



Important note to C developers:
-------------------------------

Note again that the C library at the core of this module can also be
used stand-alone (i.e., it contains no inter-dependencies whatsoever
with Perl).

The library itself consists of three files: "DateCalc.c", "DateCalc.h"
and "ToolBox.h".

Just compile "DateCalc.c" (which automatically includes "ToolBox.h")
and link the resulting output file "DateCalc.o" with your application,
which in turn should include "ToolBox.h" and "DateCalc.h" (in this order).


Example applications:
---------------------

Please refer to the file "EXAMPLES" in this distribution for details
about the example applications in the "examples/" subdirectory.


Tools:
------

Please refer to the file "TOOLS" in this distribution for details
about the various tools to be found in the "tools/" subdirectory.


Credits:
--------

Please refer to the file "CREDITS" in this distribution for a list of
contributors.


Author's note:
--------------

If you have any questions, suggestions or need any assistance, please
let me know!

I would be glad to receive any kind of feedback from you!

I hope you will find this module beneficial.

Yours,
--
  Steffen Beyer <sb@engelschall.com> http://www.engelschall.com/u/sb/
       "There is enough for the need of everyone in this world,
         but not for the greed of everyone." - Mahatma Gandhi


CHANGES.txt
                     ====================================
                       Package "Date::Calc" Version 6.2
                     ====================================


                  Copyright (c) 1995 - 2009 by Steffen Beyer.
                             All rights reserved.


Version history:
----------------

Version 6.2   16.10.2009

 +  Split the "Date::Calc" distribution into a pure-Perl and a C/XS part

Version 6.1   15.10.2009

 +  United "Date::Calc" and "Date::Pcalc" into a single distribution
 +  Fixed Polish names of months and days of week (RT ticket #14159)

Version 6.0   07.10.2009

 +  Added new functions "N_Delta_YMDHMS()", "Add_N_Delta_YMD()"
    and "Add_N_Delta_YMDHMS()" to "Date::Calc"
 +  Added more tests to "t/f037.t" for these new functions
 +  Added a new "normalized" mode to "Date::Calc::Object" which
    uses the new functions "N_Delta_YMD()", "N_Delta_YMDHMS()",
    "Add_N_Delta_YMD()" and "Add_N_Delta_YMDHMS()"
 +  Added test scripts "t/m012.t" and "t/m013.t" for this new mode
 +  The language can now be set individually for each function
    in "Date::Calc" that requires it (through a new optional
    parameter; the default continues to be a global setting
    for backward compatibility); the affected functions are:
    "Decode_Month()", "Decode_Day_of_Week()", "Compressed_to_Text()",
    "Date_to_Text()", "Date_to_Text_Long()", "Calendar()",
    "Month_to_Text()", "Day_of_Week_to_Text()", "Day_of_Week_Abbreviation()",
    "Decode_Date_EU()", "Decode_Date_US()", "Decode_Date_EU2()",
    "Decode_Date_US2()", "Parse_Date()".
 +  BEWARE that the interface of "DateCalc.c" has changed!
 +  Module "Date::Calc::Object" has been changed similarly
 +  Module "Date::Calendar::Year" has also been adapted accordingly
 +  Many test scripts have been changed to reflect the modifications
    in "Date::Calc", "Date::Calc::Object" and "Date::Calendar::Year"
    and more test cases have been added
 +  Updated the documentation to reflect all changes
 +  Updated version numbers of dependencies in "t/f000.t"

Version 5.8   12.09.2009

 +  Added a new function "N_Delta_YMD()"
 +  Added a new test script "t/f037.t"
 +  Updated the "Calc.pod" manual page accordingly
 +  Renamed "calendar.cgi" in the "examples"
    subdirectory to "calendar.pl"
 +  Added a new CGI script named "datecalc.pl"
    to the "examples" subdirectory

Version 5.7   23.08.2009

 +  Fixed "october" => "oktober" in Dutch
 +  Disabled the special abbreviated names
    of the days of the week in Portuguese
 +  Made the days which form the weekend con-
    figurable in "Calendar.pm" and "Year.pm"
 +  Added some test cases for this new feature
    in "t/m008.t"
 +  The file "examples/calendar.cgi" now
    also supports this new feature
 +  Updated "README.txt" and "INSTALL.txt"
    and the dependency on "Bit::Vector" 7.0
 +  Added an additional "README.htm" file to this
    distribution highlighting its key points
 +  Updated the documentation of "Date::Calc",
    "Date::Calendar", "Date::Calendar::Profiles",
    "Date::Calendar::Year" and "Date::Object"
 +  See the new module "Date::Calc::Util" (which
    is available separately; to be released soon)
    for all the shortcuts you ever wanted to have
    in "Date::Calc"

Version 5.6   28.07.2009

 +  Made the module MacOS X compatible
 +  Made some tiny changes to the documentation

Version 5.5   was skipped due to an unauthorized upload by someone else

Version 5.4   03.10.2004

 +  Added compiler directives for C++.
 +  Removed "Carp::Clan" from the distribution (available separately).
 +  Fixed bug in initialization of "Date::Calendar::Year" objects.
 +  Added method "tags()" to "Date::Calendar" and "Date::Calendar::Year".
 +  Fixed the formula for "Labor Day" in the U.S. to "1/Mon/Sep".
 +  Added a new recipe to the "Date::Calc" documentation.
 +  Added Romanian to the list of languages supported by "Date::Calc".
 +  Changed the example script "calendar.cgi" to highlight the name
    which led to a given date being a holiday.
 +  Fixed the Polish entries in "Date::Calc".
 +  Added a few commemorative days to the Norwegian calendar profile.
 +  Added "use bytes" to all Perl files to avoid problems on systems
    not using the standard locale "C".
 +  Fixed test 5 of t/m005.t to (hopefully) work under other locales.

Version 5.3   29.09.2002

 +  Simplified the error message handlers in "Calc.xs".

Version 5.2   18.09.2002

 +  Changed the Polish profiles from ISO-Latin-2 to ISO-Latin-1.
 +  Fixed the broken tests in "t/f034.t" and "t/m006.t".
 +  Synchronized "Carp::Clan" and "ToolBox.h" with "Bit::Vector" 6.2.

Version 5.1   08.09.2002

 +  Integrated modifications needed for MacOS / MacPerl.
 +  Added new method "normalize()" (Date::Calc::Object).
 +  Added a new test script "t/m011.t" for "normalize()".
 +  Added a calendar profile for Mexico (Date::Calendar::Profiles).
 +  Little additions to and corrections of the documentation (Date::Calc).
 +  Added Polish (Date::Calc) and a calendar profile for Poland.
 +  Added Hungarian (Date::Calc).
 +  Added some more commemorative days to the profile of Brazil.

Version 5.0   10.10.2001

Module "Date::Calc":

 +  Added the following new functions:
     *  check_time()
     *  Delta_YMD()
     *  Delta_YMDHMS()
     *  Add_Delta_YM()
     *  Add_Delta_YMDHMS()
     *  Normalize_DHMS()
     *  This_Year()
     *  Gmtime()
     *  Localtime()
     *  Mktime()
     *  Timezone()
     *  Date_to_Time()
     *  Time_to_Date()
     *  Fixed_Window()
     *  Moving_Window()
     *  ISO_LC()
     *  ISO_UC()
 +  Added support for Norwegian.
 +  Added support for Swedish.
 +  Added support for Danish.
 +  Added support for Finnish.
 +  Changed the month names of some languages to lower case.
 +  Changed the french "long" date format to a more popular form.
 +  Changed the corresponding test script (t/f012.t) accordingly.
 +  Corrected the spanish texts in ./examples/age_in_days_*.

ToolBox.h
#ifndef MODULE_TOOLBOX
#define MODULE_TOOLBOX
#ifdef __cplusplus
extern "C"
{
#endif
/*****************************************************************************/
/*  MODULE NAME:  ToolBox.h                             MODULE TYPE:  (dat)  */
/*****************************************************************************/
/*  MODULE IMPORTS:                                                          */
/*****************************************************************************/

/*****************************************************************************/
/*  MODULE INTERFACE:                                                        */
/*****************************************************************************/

/*****************************************************************************/
/*  MODULE RESOURCES:                                                        */
/*****************************************************************************/

/*****************************************************************************/
/*  NOTE: The type names that have been chosen here are somewhat weird on    */
/*        purpose, in order to avoid name clashes with system header files   */
/*        and your own application(s) which might - directly or indirectly - */
/*        include this definitions file.                                     */
/*****************************************************************************/

typedef  unsigned   char    N_char;
typedef  unsigned   char    N_byte;
typedef  unsigned   short   N_short;
typedef  unsigned   short   N_shortword;
typedef  unsigned   int     N_int;
typedef  unsigned   int     N_word;
typedef  unsigned   long    N_long;
typedef  unsigned   long    N_longword;

/*  Mnemonic 1:  The natural numbers,  N = { 0, 1, 2, 3, ... }               */
/*  Mnemonic 2:  Nnnn = u_N_signed,  _N_ot signed                            */

typedef  signed     char    Z_char;
typedef  signed     char    Z_byte;
typedef  signed     short   Z_short;
typedef  signed     short   Z_shortword;
typedef  signed     int     Z_int;
typedef  signed     int     Z_word;
typedef  signed     long    Z_long;
typedef  signed     long    Z_longword;

/*  Mnemonic 1:  The whole numbers,  Z = { 0, -1, 1, -2, 2, -3, 3, ... }     */
/*  Mnemonic 2:  Zzzz = Ssss_igned                                           */

typedef  void               *voidptr;
typedef  N_char             *charptr;
typedef  N_byte             *byteptr;
typedef  N_short            *shortptr;
typedef  N_shortword        *shortwordptr;
typedef  N_int              *intptr;
typedef  N_word             *wordptr;
typedef  N_long             *longptr;
typedef  N_longword         *longwordptr;

typedef  N_char             *N_charptr;
typedef  N_byte             *N_byteptr;
typedef  N_short            *N_shortptr;
typedef  N_shortword        *N_shortwordptr;
typedef  N_int              *N_intptr;
typedef  N_word             *N_wordptr;
typedef  N_long             *N_longptr;
typedef  N_longword         *N_longwordptr;

typedef  Z_char             *Z_charptr;
typedef  Z_byte             *Z_byteptr;
typedef  Z_short            *Z_shortptr;
typedef  Z_shortword        *Z_shortwordptr;
typedef  Z_int              *Z_intptr;
typedef  Z_word             *Z_wordptr;
typedef  Z_long             *Z_longptr;
typedef  Z_longword         *Z_longwordptr;

#undef  FALSE
#define FALSE       (0!=0)

#undef  TRUE
#define TRUE        (0==0)

#ifdef __cplusplus
    typedef bool boolean;
#else
    #ifdef MACOS_TRADITIONAL
        #define boolean Boolean
    #else
        typedef enum { false = FALSE, true = TRUE } boolean;
    #endif
#endif

#define and         &&      /* logical (boolean) operators: lower case */
#define or          ||
#define not         !

#define AND         &       /* binary (bitwise) operators: UPPER CASE */
#define OR          |
#define XOR         ^
#define NOT         ~
#define SHL         <<
#define SHR         >>

#ifdef ENABLE_MODULO
#define mod         %       /* arithmetic operators */
#endif

#define blockdef(name,size)         unsigned char name[size]
#define blocktypedef(name,size)     typedef unsigned char name[size]

/*******************************************************************************/
/*  MODULE IMPLEMENTATION:                                                     */
/*******************************************************************************/

/*******************************************************************************/
/*  VERSION:  5.5.1                                                            */
/*******************************************************************************/
/*  VERSION HISTORY:                                                           */
/*******************************************************************************/
/*                                                                             */
/*    Version 5.5.1  28.02.07  Uploaded correct version to CPAN.               */
/*    Version 5.5    23.01.07  Fixed compiler warning under gcc-4.             */
/*    Version 5.4    03.10.04  Added compiler directives for C++.              */
/*    Version 5.3    29.09.02  No changes.                                     */
/*    Version 5.2    18.09.02  No changes.                                     */
/*    Version 5.1    08.09.02  Added conditional changes for MacOS/MacPerl.    */
/*    Version 5.0    10.10.01  New YMD/HMS functions, replaced <ctype.h>, ...  */
/*    Version 4.3    08.01.00  decode_date_??: (yy < 70 ? 20yy : 19yy)         */
/*    Version 4.2    07.09.98  No changes.                                     */
/*    Version 4.1    08.06.98  Fixed bug in "add_delta_ymd()".                 */
/*    Version 4.0    12.05.98  Major rework. Added multi-language support.     */
/*    Version 3.2    15.06.97  Added "week_of_year()".                         */
/*    Version 3.1    12.06.97  No significant changes.                         */
/*    Version 3.0    16.02.97  Changed conventions for unsuccessful returns.   */
/*    Version 2.3    22.11.96  Fixed unbalanced "malloc" and "free".           */
/*    Version 2.2    26.05.96  No significant changes.                         */
/*    Version 2.1    26.05.96  Fixed HH MM SS parameter checks.                */
/*    Version 2.0    25.05.96  Added time calculations. Major rework.          */
/*    Version 1.6    20.04.96  Not published.                                  */
/*    Version 1.5    14.03.96  No significant changes.                         */
/*    Version 1.4    11.02.96  No significant changes.                         */
/*    Version 1.3    10.12.95  Added "days_in_month()".                        */
/*    Version 1.2b   27.11.95  No significant changes.                         */
/*    Version 1.2a   21.11.95  Fix for type name clashes.                      */
/*    Version 1.1    18.11.95  Fix for type name clashes.                      */
/*    Version 1.01   16.11.95  Improved compliance w/ programming standards.   */
/*    Version 1.0    14.11.95  First version under UNIX (with Perl module).    */
/*    Version 0.9    01.11.93  First version of C library under MS-DOS.        */
/*                                                                             */
/*******************************************************************************/
/*  AUTHOR:                                                                    */
/*******************************************************************************/
/*                                                                             */
/*    Steffen Beyer                                                            */
/*    mailto:sb@engelschall.com                                                */
/*    http://www.engelschall.com/u/sb/download/                                */
/*                                                                             */
/*******************************************************************************/
/*  COPYRIGHT:                                                                 */
/*******************************************************************************/
/*                                                                             */
/*    Copyright (c) 1993 - 2004 by Steffen Beyer.                              */
/*    All rights reserved.                                                     */
/*                                                                             */
/*******************************************************************************/
/*  LICENSE:                                                                   */
/*******************************************************************************/
/*                                                                             */
/*    This library is free software; you can redistribute it and/or            */
/*    modify it under the terms of the GNU Library General Public              */
/*    License as published by the Free Software Foundation; either             */
/*    version 2 of the License, or (at your option) any later version.         */
/*                                                                             */
/*    This library is distributed in the hope that it will be useful,          */
/*    but WITHOUT ANY WARRANTY; without even the implied warranty of           */
/*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU         */
/*    Library General Public License for more details.                         */
/*                                                                             */
/*    You should have received a copy of the GNU Library General Public        */
/*    License along with this library; if not, write to the                    */
/*    Free Software Foundation, Inc.,                                          */
/*    59 Temple Place, Suite 330, Boston, MA 02111-1307 USA                    */
/*    or download a copy from ftp://ftp.gnu.org/pub/gnu/COPYING.LIB-2.0        */
/*                                                                             */
/*******************************************************************************/
#ifdef __cplusplus
}
#endif
#endif
